column.replace() may fail if the resulting string column is too big

```
>>> DT = dt.Frame(a=["a"] * 10000000)
>>> DT.replace("a", "A" * 250)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Assertion 'sb.size() == (mb.get_element<T>(n) & ~GETNA<T>())' failed in "c/column_string.cc", line 38
```
The problem here is that the size of the column created exceeds the limits of `int32`. The proper course of action in this case is to create the STR64 column instead.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -57,6 +57,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - A keyed frame will now be rendered correctly when viewing it in python
   console via `Frame.view()` (#1672).
 
+- Str32 column can no longer overflow during the `.replace()` operation,
+  or when converting from python, numpy or pandas, etc. In all these cases
+  we will now transparently create a Str64 column instead (#1694).
+
 
 ### Changed
 
@@ -86,7 +90,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
   [arno candel][] (#1619),
   [antorsae][] (#1639),
-  [pasha stetsenko][] (#1672)
+  [pasha stetsenko][] (#1672, #1694)
 
 
 

--- a/c/column.cc
+++ b/c/column.cc
@@ -127,19 +127,6 @@ Column* Column::new_mbuf_column(SType stype, MemoryRange&& mbuf) {
   return col;
 }
 
-Column* Column::new_mbuf_column(SType stype, MemoryRange&& mbuf,
-                                MemoryRange&& strbuf)
-{
-  Column* col = new_column(stype);
-  if (stype == SType::STR32 || stype == SType::STR64) {
-    col->replace_buffer(std::move(mbuf), std::move(strbuf));
-  } else {
-    xassert(!strbuf);
-    col->replace_buffer(std::move(mbuf));
-  }
-  return col;
-}
-
 
 
 void Column::replace_buffer(MemoryRange&&) {

--- a/c/column.h
+++ b/c/column.h
@@ -588,7 +588,6 @@ template <typename T> class StringColumn : public Column
 
 public:
   StringColumn(size_t nrows);
-  StringColumn(size_t nrows, MemoryRange&& offbuf, MemoryRange&& strbuf);
   void save_to_disk(const std::string& filename,
                     WritableBuffer::Strategy strategy) override;
   void replace_buffer(MemoryRange&&, MemoryRange&&) override;
@@ -623,6 +622,7 @@ public:
 
 protected:
   StringColumn();
+  StringColumn(size_t nrows, MemoryRange&& offbuf, MemoryRange&& strbuf);
   void init_data() override;
   void init_mmap(const std::string& filename) override;
   void open_mmap(const std::string& filename, bool recode) override;
@@ -647,8 +647,11 @@ protected:
 
   friend Column;
   friend FreadReader;  // friend Column* alloc_column(SType, size_t, int);
+  friend Column* new_string_column(size_t, MemoryRange&&, MemoryRange&&);
 };
 
+
+Column* new_string_column(size_t n, MemoryRange&& data, MemoryRange&& str);
 
 template <> void StringColumn<uint32_t>::cast_into(StringColumn<uint64_t>*) const;
 template <> void StringColumn<uint64_t>::cast_into(StringColumn<uint64_t>*) const;

--- a/c/column.h
+++ b/c/column.h
@@ -118,6 +118,10 @@ public:  // TODO: convert this into private
   size_t nrows;
 
 public:
+  static constexpr size_t MAX_STRING_SIZE = 0x7FFFFFFF;
+  static constexpr size_t MAX_STR32_BUFFER_SIZE = 0x7FFFFFFF;
+  static constexpr size_t MAX_STR32_NROWS = 0x7FFFFFFF;
+
   static Column* new_data_column(SType, size_t nrows);
   static Column* new_na_column(SType, size_t nrows);
   static Column* new_mmap_column(SType, size_t nrows, const std::string& filename);
@@ -125,7 +129,6 @@ public:
                                   bool recode = false);
   static Column* new_xbuf_column(SType, size_t nrows, Py_buffer* pybuffer);
   static Column* new_mbuf_column(SType, MemoryRange&&);
-  static Column* new_mbuf_column(SType, MemoryRange&&, MemoryRange&&);
   static Column* from_pylist(const py::olist& list, int stype0 = 0);
   static Column* from_pylist_of_tuples(const py::olist& list, size_t index, int stype0);
   static Column* from_pylist_of_dicts(const py::olist& list, py::robj name, int stype0);
@@ -643,15 +646,19 @@ protected:
   void cast_into(StringColumn<uint64_t>*) const override;
   void fill_na() override;
 
-  //int verify_meta_integrity(std::vector<char>*, int, const char* = "Column") const override;
-
   friend Column;
   friend FreadReader;  // friend Column* alloc_column(SType, size_t, int);
   friend Column* new_string_column(size_t, MemoryRange&&, MemoryRange&&);
 };
 
 
+/**
+ * Use this function to create a column from existing offsets & strdata.
+ * It will create either StringColumn<uint32_t>* or StringColumn<uint64_t>*
+ * depending on the size of the data.
+ */
 Column* new_string_column(size_t n, MemoryRange&& data, MemoryRange&& str);
+
 
 template <> void StringColumn<uint32_t>::cast_into(StringColumn<uint64_t>*) const;
 template <> void StringColumn<uint64_t>::cast_into(StringColumn<uint64_t>*) const;

--- a/c/column_bool.cc
+++ b/c/column_bool.cc
@@ -75,7 +75,7 @@ inline static MemoryRange cast_str_helper(
   for (size_t i = 0; i < src->nrows; ++i) {
     int8_t x = src_data[i];
     if (ISNA<int8_t>(x)) {
-      toffsets[i] = offset | GETNA<T>();
+      toffsets[i] = offset ^ GETNA<T>();
     } else {
       *ch++ = '0' + x;
       offset++;

--- a/c/column_from_pylist.cc
+++ b/c/column_from_pylist.cc
@@ -480,17 +480,18 @@ static bool parse_as_pyobj(const iterable* list, MemoryRange& membuf)
 // Parse controller
 //------------------------------------------------------------------------------
 
-static int find_next_stype(int curr_stype, int stype0) {
+static SType find_next_stype(SType curr_stype, int stype0) {
+  int istype = static_cast<int>(curr_stype);
   if (stype0 > 0) {
-    return stype0;
+    return static_cast<SType>(stype0);
   }
   if (stype0 < 0) {
-    return std::min(curr_stype + 1, -stype0);
+    return static_cast<SType>(std::min(istype + 1, -stype0));
   }
-  if (curr_stype == DT_STYPES_COUNT - 1) {
+  if (istype == DT_STYPES_COUNT - 1) {
     return curr_stype;
   }
-  return (curr_stype + 1) % int(DT_STYPES_COUNT);
+  return static_cast<SType>((istype + 1) % int(DT_STYPES_COUNT));
 }
 
 
@@ -500,12 +501,12 @@ Column* Column::from_py_iterable(const iterable* il, int stype0)
   MemoryRange membuf;
   MemoryRange strbuf;
   // TODO: Perhaps `stype` and `curr_stype` should have type SType ?
-  int stype = find_next_stype(0, stype0);
+  SType stype = find_next_stype(SType::VOID, stype0);
   size_t i = 0;
-  while (stype) {
-    int next_stype = find_next_stype(stype, stype0);
+  while (stype != SType::VOID) {
+    SType next_stype = find_next_stype(stype, stype0);
     if (stype == next_stype) {
-      switch (static_cast<SType>(stype)) {
+      switch (stype) {
         case SType::BOOL:    force_as_bool(il, membuf); break;
         case SType::INT8:    force_as_int<int8_t>(il, membuf); break;
         case SType::INT16:   force_as_int<int16_t>(il, membuf); break;
@@ -523,7 +524,7 @@ Column* Column::from_py_iterable(const iterable* il, int stype0)
       break; // while(stype)
     } else {
       bool ret = false;
-      switch (static_cast<SType>(stype)) {
+      switch (stype) {
         case SType::BOOL:    ret = parse_as_bool(il, membuf, i); break;
         case SType::INT8:    ret = parse_as_int<int8_t>(il, membuf, i); break;
         case SType::INT16:   ret = parse_as_int<int16_t>(il, membuf, i); break;
@@ -539,18 +540,16 @@ Column* Column::from_py_iterable(const iterable* il, int stype0)
       stype = next_stype;
     }
   }
-  Column* col = Column::new_column(static_cast<SType>(stype));
-  if (static_cast<SType>(stype) == SType::OBJ) {
-    membuf.set_pyobjects(/* clear_data = */ false);
+  if (stype == SType::STR32 || stype == SType::STR64) {
+    size_t nrows = il->size();
+    return new_string_column(nrows, std::move(membuf), std::move(strbuf));
   }
-  if (static_cast<SType>(stype) == SType::STR32 ||
-      static_cast<SType>(stype) == SType::STR64)
-  {
-    col->replace_buffer(std::move(membuf), std::move(strbuf));
-  } else {
-    col->replace_buffer(std::move(membuf));
+  else {
+    if (stype == SType::OBJ) {
+      membuf.set_pyobjects(/* clear_data = */ false);
+    }
+    return Column::new_mbuf_column(stype, std::move(membuf));
   }
-  return col;
 }
 
 

--- a/c/column_from_pylist.cc
+++ b/c/column_from_pylist.cc
@@ -335,7 +335,7 @@ static bool parse_as_str(const iterable* list, MemoryRange& offbuf,
     py::robj item = list->item(i);
 
     if (item.is_none()) {
-      offsets[i] = curr_offset | GETNA<T>();
+      offsets[i] = curr_offset ^ GETNA<T>();
       continue;
     }
     if (item.is_string()) {
@@ -411,7 +411,7 @@ static void force_as_str(const iterable* list, MemoryRange& offbuf,
     py::oobj item = list->item(i);
 
     if (item.is_none()) {
-      offsets[i] = curr_offset | GETNA<T>();
+      offsets[i] = curr_offset ^ GETNA<T>();
       continue;
     }
     if (!item.is_string()) {
@@ -425,7 +425,7 @@ static void force_as_str(const iterable* list, MemoryRange& offbuf,
         if (std::is_same<T, int32_t>::value &&
               (static_cast<int64_t>(tlen) != cstr.size ||
                next_offset < curr_offset)) {
-          offsets[i] = curr_offset | GETNA<T>();
+          offsets[i] = curr_offset ^ GETNA<T>();
           continue;
         }
         if (strbuf.size() < static_cast<size_t>(next_offset)) {
@@ -441,7 +441,7 @@ static void force_as_str(const iterable* list, MemoryRange& offbuf,
       offsets[i] = curr_offset;
       continue;
     } else {
-      offsets[i] = curr_offset | GETNA<T>();
+      offsets[i] = curr_offset ^ GETNA<T>();
     }
   }
   strbuf.resize(curr_offset);

--- a/c/column_int.cc
+++ b/c/column_int.cc
@@ -95,7 +95,7 @@ inline static MemoryRange cast_str_helper(
   for (size_t i = 0; i < nrows; ++i) {
     IT x = src[i];
     if (ISNA<IT>(x)) {
-      toffsets[i] = offset | GETNA<OT>();
+      toffsets[i] = offset ^ GETNA<OT>();
     } else {
       char* ch0 = ch;
       toa<IT>(&ch, x);

--- a/c/column_pyobj.cc
+++ b/c/column_pyobj.cc
@@ -125,7 +125,7 @@ inline static MemoryRange cast_str_helper(
       offset += static_cast<OT>(xcstr.size);
       toffsets[i] = offset;
     } else {
-      toffsets[i] = offset | GETNA<OT>();
+      toffsets[i] = offset ^ GETNA<OT>();
     }
   }
   wb->finalize();

--- a/c/column_real.cc
+++ b/c/column_real.cc
@@ -81,7 +81,7 @@ inline static MemoryRange cast_str_helper(
   for (size_t i = 0; i < src->nrows; ++i) {
     IT x = src_data[i];
     if (ISNA<IT>(x)) {
-      toffsets[i] = offset | GETNA<OT>();
+      toffsets[i] = offset ^ GETNA<OT>();
     } else {
       char* ch0 = ch;
       toa<IT>(&ch, x);

--- a/c/column_string.cc
+++ b/c/column_string.cc
@@ -20,42 +20,75 @@
 static std::string path_str(const std::string& path);
 
 
+//------------------------------------------------------------------------------
+// String column construction
+//------------------------------------------------------------------------------
+
+// public: create a string column for `n` rows, preallocating the offsets array
+// but leaving string buffer empty (and not allocated).
+template <typename T>
+StringColumn<T>::StringColumn(size_t n) : Column(n)
+{
+  mbuf = MemoryRange::mem(sizeof(T) * (n + 1));
+  mbuf.set_element<T>(0, 0);
+}
+
+
+// private use only
 template <typename T>
 StringColumn<T>::StringColumn() : Column(0) {}
 
 
-template <typename T>
-StringColumn<T>::StringColumn(size_t n)
-  : StringColumn<T>(n, MemoryRange(), MemoryRange()) {}
-
-
+// private: use `new_string_column(n, &&mb, &&sb)` instead
 template <typename T>
 StringColumn<T>::StringColumn(size_t n, MemoryRange&& mb, MemoryRange&& sb)
   : Column(n)
 {
-  size_t expected_offsets_size = sizeof(T) * (n + 1);
-  if (mb) {
-    xassert(mb.size() == expected_offsets_size);
-    xassert(mb.get_element<T>(0) == 0);
-    xassert(sb.size() == (mb.get_element<T>(n) & ~GETNA<T>()));
-    mbuf = std::move(mb);
-    strbuf = std::move(sb);
-  } else {
-    xassert(!sb);
-    mbuf = MemoryRange::mem(expected_offsets_size);
-    mbuf.set_element<T>(0, 0);
+  xassert(mb);
+  xassert(mb.size() == sizeof(T) * (n + 1));
+  xassert(mb.get_element<T>(0) == 0);
+  xassert(sb.size() == (mb.get_element<T>(n) & ~GETNA<T>()));
+  mbuf = std::move(mb);
+  strbuf = std::move(sb);
+}
+
+
+static MemoryRange _recode_offsets_to_u64(const MemoryRange& offsets) {
+  // TODO: make this parallel
+  MemoryRange off64 = MemoryRange::mem(offsets.size() * 2);
+  auto data64 = static_cast<uint64_t*>(off64.xptr());
+  auto data32 = static_cast<const uint32_t*>(offsets.rptr());
+  data64[0] = 0;
+  uint64_t curr_offset = 0;
+  size_t n = offsets.size() / sizeof(uint32_t) - 1;
+  for (size_t i = 1; i <= n; ++i) {
+    uint32_t len = data32[i] - data32[i - 1];
+    if (len == GETNA<uint32_t>()) {
+      data64[i] = curr_offset | GETNA<uint64_t>();
+    } else {
+      curr_offset += len;
+      data64[i] = curr_offset;
+    }
   }
+  return off64;
 }
 
 
 Column* new_string_column(size_t n, MemoryRange&& data, MemoryRange&& str) {
-  if (data.size() == sizeof(uint32_t) * (n + 1)) {
-    return new StringColumn<uint32_t>(n, std::move(data), std::move(str));
+  size_t data_size = data.size();
+  size_t strb_size = str.size();
+
+  if (data_size == sizeof(uint32_t) * (n + 1)) {
+    if (strb_size <= Column::MAX_STR32_BUFFER_SIZE &&
+        n <= Column::MAX_STR32_NROWS) {
+      return new StringColumn<uint32_t>(n, std::move(data), std::move(str));
+    }
+    // Otherwise, offsets need to be recoded into a uint64_t array
+    data = _recode_offsets_to_u64(data);
   }
-  else {
-    return new StringColumn<uint64_t>(n, std::move(data), std::move(str));
-  }
+  return new StringColumn<uint64_t>(n, std::move(data), std::move(str));
 }
+
 
 
 
@@ -87,13 +120,6 @@ void StringColumn<T>::open_mmap(const std::string& filename, bool recode) {
   mbuf = MemoryRange::mmap(filename);
   strbuf = MemoryRange::mmap(filename_str);
 
-  // size_t exp_mbuf_size = sizeof(T) * (nrows + 1);
-  // if (mbuf.size() != exp_mbuf_size) {
-  //   throw Error() << "File \"" << filename <<
-  //       "\" cannot be used to create a column with " << nrows <<
-  //       " rows. Expected file size of " << exp_mbuf_size <<
-  //       " bytes, actual size is " << mbuf.size() << " bytes";
-  // }
   if (recode) {
     if (mbuf.get_element<T>(0) != 0) {
       // Recode old format of string storage
@@ -105,15 +131,6 @@ void StringColumn<T>::open_mmap(const std::string& filename, bool recode) {
       }
     }
   }
-
-  // size_t exp_strbuf_size = (mbuf.get_element<T>(nrows) & ~GETNA<T>());
-  // if (strbuf.size() != exp_strbuf_size) {
-  //   size_t strbuf_size = strbuf.size();
-  //   throw Error() << "File \"" << filename_str <<
-  //     "\" cannot be used to create a column with " << nrows <<
-  //     " rows. Expected file size of " << exp_strbuf_size <<
-  //     " bytes, actual size is " << strbuf_size << " bytes";
-  // }
 }
 
 // Not implemented (should it be?) see method signature in `Column` for

--- a/c/column_string.cc
+++ b/c/column_string.cc
@@ -64,9 +64,9 @@ static MemoryRange _recode_offsets_to_u64(const MemoryRange& offsets) {
   for (size_t i = 1; i <= n; ++i) {
     uint32_t len = data32[i] - data32[i - 1];
     if (len == GETNA<uint32_t>()) {
-      data64[i] = curr_offset | GETNA<uint64_t>();
+      data64[i] = curr_offset ^ GETNA<uint64_t>();
     } else {
-      curr_offset += len;
+      curr_offset += len & ~GETNA<uint32_t>();
       data64[i] = curr_offset;
     }
   }
@@ -297,7 +297,7 @@ void StringColumn<T>::reify() {
     size_t j = start;
     for (size_t i = 0; i < nrows; ++i, j += step) {
       if (ISNA<T>(offs1[j])) {
-        offs_dest[i] = prev_off | GETNA<T>();
+        offs_dest[i] = prev_off ^ GETNA<T>();
       } else {
         T off0 = offs0[j] & ~GETNA<T>();
         T str_len = offs1[j] - off0;
@@ -331,7 +331,7 @@ void StringColumn<T>::reify() {
     ri.iterate(0, nrows, 1,
       [&](size_t i, size_t j) {
         if (j == RowIndex::NA || ISNA<T>(offs1[j])) {
-          offs_dest[i] = prev_off | GETNA<T>();
+          offs_dest[i] = prev_off ^ GETNA<T>();
         } else {
           T off0 = offs0[j] & ~GETNA<T>();
           T str_len = offs1[j] - off0;
@@ -457,7 +457,7 @@ void StringColumn<T>::resize_and_fill(size_t new_nrows)
       strbuf = new_strbuf;
     } else {
       if (old_nrows == 1) xassert(old_strbuf_size == 0);
-      T na = static_cast<T>(old_strbuf_size) | GETNA<T>();
+      T na = static_cast<T>(old_strbuf_size) ^ GETNA<T>();
       set_value(offsets + nrows, &na, sizeof(T), new_nrows - old_nrows);
     }
   }
@@ -482,7 +482,7 @@ void StringColumn<T>::apply_na_mask(const BoolColumn* mask) {
     T offa = offi & ~GETNA<T>();
     if (maskdata[j] == 1) {
       doffset += offa - offp;
-      offsets[j] = offp | GETNA<T>();
+      offsets[j] = offp ^ GETNA<T>();
       continue;
     } else if (doffset) {
       if (offi == offa) {
@@ -490,7 +490,7 @@ void StringColumn<T>::apply_na_mask(const BoolColumn* mask) {
         std::memmove(strdata + offp, strdata + offp + doffset,
                      offi - offp - doffset);
       } else {
-        offsets[j] = offp | GETNA<T>();
+        offsets[j] = offp ^ GETNA<T>();
       }
     }
     offp = offa;
@@ -708,7 +708,7 @@ void StringColumn<T>::verify_integrity(const std::string& name) const {
   for (size_t i = 0; i < mbuf_nrows; ++i) {
     T oj = str_offsets[i];
     if (ISNA<T>(oj)) {
-      if (oj != (lastoff | GETNA<T>())) {
+      if (oj != (lastoff ^ GETNA<T>())) {
         throw AssertionError()
             << "Offset of NA String in row " << i << " of " << name
             << " does not have the same magnitude as the previous offset: "

--- a/c/column_string.cc
+++ b/c/column_string.cc
@@ -19,32 +19,44 @@
 // the path to the offsets
 static std::string path_str(const std::string& path);
 
+
 template <typename T>
 StringColumn<T>::StringColumn() : Column(0) {}
 
+
 template <typename T>
-StringColumn<T>::StringColumn(size_t nrows_)
-  : StringColumn<T>(nrows_, MemoryRange(), MemoryRange()) {}
+StringColumn<T>::StringColumn(size_t n)
+  : StringColumn<T>(n, MemoryRange(), MemoryRange()) {}
 
 
 template <typename T>
 StringColumn<T>::StringColumn(size_t n, MemoryRange&& mb, MemoryRange&& sb)
   : Column(n)
 {
-  size_t exp_off_size = sizeof(T) * (n + 1);
+  size_t expected_offsets_size = sizeof(T) * (n + 1);
   if (mb) {
-    xassert(mb.size() == exp_off_size);
+    xassert(mb.size() == expected_offsets_size);
     xassert(mb.get_element<T>(0) == 0);
     xassert(sb.size() == (mb.get_element<T>(n) & ~GETNA<T>()));
+    mbuf = std::move(mb);
+    strbuf = std::move(sb);
   } else {
     xassert(!sb);
-    mb = MemoryRange::mem(exp_off_size);
-    mb.set_element<T>(0, 0);
+    mbuf = MemoryRange::mem(expected_offsets_size);
+    mbuf.set_element<T>(0, 0);
   }
-
-  mbuf = std::move(mb);
-  strbuf = std::move(sb);
 }
+
+
+Column* new_string_column(size_t n, MemoryRange&& data, MemoryRange&& str) {
+  if (data.size() == sizeof(uint32_t) * (n + 1)) {
+    return new StringColumn<uint32_t>(n, std::move(data), std::move(str));
+  }
+  else {
+    return new StringColumn<uint64_t>(n, std::move(data), std::move(str));
+  }
+}
+
 
 
 //==============================================================================

--- a/c/csv/reader.cc
+++ b/c/csv/reader.cc
@@ -789,6 +789,7 @@ void GenericReader::report_columns_to_python() {
 
 dtptr GenericReader::makeDatatable() {
   size_t ncols = columns.size();
+  size_t nrows = columns.get_nrows();
   size_t ocols = columns.nColumnsInOutput();
   std::vector<Column*> ccols;
   ccols.reserve(ocols);
@@ -797,8 +798,11 @@ dtptr GenericReader::makeDatatable() {
     if (!col.is_in_output()) continue;
     MemoryRange databuf = col.extract_databuf();
     MemoryRange strbuf = col.extract_strbuf();
-    ccols.push_back(Column::new_mbuf_column(col.get_stype(), std::move(databuf),
-                                       std::move(strbuf)));
+    SType stype = col.get_stype();
+    ccols.push_back((stype == SType::STR32 || stype == SType::STR64)
+      ? new_string_column(nrows, std::move(databuf), std::move(strbuf))
+      : Column::new_mbuf_column(stype, std::move(databuf))
+    );
   }
   py::olist names = freader.get_attr("_colnames").to_pylist();
   return dtptr(new DataTable(std::move(ccols), names));

--- a/c/expr/repl_node.cc
+++ b/c/expr/repl_node.cc
@@ -366,12 +366,7 @@ colptr scalar_string_rn::make_column(SType st, size_t nrows) const {
   }
   MemoryRange strbuf = MemoryRange::mem(len);
   std::memcpy(strbuf.xptr(), value.data(), len);
-  Column* col = nullptr;
-  if (elemsize == 4) {
-    col = new StringColumn<uint32_t>(1, std::move(offbuf), std::move(strbuf));
-  } else {
-    col = new StringColumn<uint64_t>(1, std::move(offbuf), std::move(strbuf));
-  }
+  Column* col = new_string_column(1, std::move(offbuf), std::move(strbuf));
   col->replace_rowindex(RowIndex(size_t(0), nrows, 0));
   return colptr(col);
 }

--- a/c/frame/rbind.cc
+++ b/c/frame/rbind.cc
@@ -394,7 +394,7 @@ void StringColumn<T>::rbind_impl(std::vector<const Column*>& columns,
       rows_to_fill += col->nrows;
     } else {
       if (rows_to_fill) {
-        const T na = curr_offset | GETNA<T>();
+        const T na = curr_offset ^ GETNA<T>();
         set_value(offs, &na, sizeof(T), rows_to_fill);
         offs += rows_to_fill;
         rows_to_fill = 0;
@@ -416,7 +416,7 @@ void StringColumn<T>::rbind_impl(std::vector<const Column*>& columns,
     delete col;
   }
   if (rows_to_fill) {
-    const T na = curr_offset | GETNA<T>();
+    const T na = curr_offset ^ GETNA<T>();
     set_value(offs, &na, sizeof(T), rows_to_fill);
   }
 }

--- a/c/frame/replace.cc
+++ b/c/frame/replace.cc
@@ -569,6 +569,7 @@ void ReplaceAgent::process_str_column(size_t colidx) {
   }
   Column* newcol = replace_str<T>(x_str.size(), x_str.data(), y_str.data(),
                                   col);
+  columns_cast = (newcol->stype() != col->stype());
   dt->columns[colidx] = newcol;
   delete col;
 }

--- a/c/frame/stats.cc
+++ b/c/frame/stats.cc
@@ -57,7 +57,7 @@ static Column* _make_column_str(CString value) {
     mbuf.set_element<T>(0, 0);
     mbuf.set_element<T>(1, GETNA<T>());
   }
-  return new StringColumn<T>(1, std::move(mbuf), std::move(strbuf));
+  return new_string_column(1, std::move(mbuf), std::move(strbuf));
 }
 
 static Column* _nacol(Stats*, const Column* col) {

--- a/c/jay/open_jay.cc
+++ b/c/jay/open_jay.cc
@@ -14,7 +14,8 @@
 
 
 // Helper functions
-static Column* column_from_jay(const jay::Column* jaycol,
+static Column* column_from_jay(size_t nrows,
+                               const jay::Column* jaycol,
                                const MemoryRange& jaybuf);
 
 
@@ -75,7 +76,7 @@ DataTable* open_jay_from_mbuf(const MemoryRange& mbuf)
   columns.reserve(ncols);
   size_t i = 0;
   for (const jay::Column* jcol : *msg_columns) {
-    Column* col = column_from_jay(jcol, mbuf);
+    Column* col = column_from_jay(nrows, jcol, mbuf);
     if (col->nrows != nrows) {
       throw IOError() << "Length of column " << i << " is " << col->nrows
           << ", however the Frame contains " << nrows << " rows";
@@ -118,29 +119,30 @@ static void initStats(Stats* stats, const jay::Column* jcol) {
 
 
 static Column* column_from_jay(
-    const jay::Column* jcol, const MemoryRange& jaybuf)
+    size_t nrows, const jay::Column* jcol, const MemoryRange& jaybuf)
 {
   jay::Type jtype = jcol->type();
 
-  Column* col = nullptr;
+  SType stype = SType::VOID;
   switch (jtype) {
-    case jay::Type_Bool8:   col = new BoolColumn(0); break;
-    case jay::Type_Int8:    col = new IntColumn<int8_t>(0); break;
-    case jay::Type_Int16:   col = new IntColumn<int16_t>(0); break;
-    case jay::Type_Int32:   col = new IntColumn<int32_t>(0); break;
-    case jay::Type_Int64:   col = new IntColumn<int64_t>(0); break;
-    case jay::Type_Float32: col = new RealColumn<float>(0); break;
-    case jay::Type_Float64: col = new RealColumn<double>(0); break;
-    case jay::Type_Str32:   col = new StringColumn<uint32_t>(0); break;
-    case jay::Type_Str64:   col = new StringColumn<uint64_t>(0); break;
+    case jay::Type_Bool8:   stype = SType::BOOL; break;
+    case jay::Type_Int8:    stype = SType::INT8; break;
+    case jay::Type_Int16:   stype = SType::INT16; break;
+    case jay::Type_Int32:   stype = SType::INT32; break;
+    case jay::Type_Int64:   stype = SType::INT64; break;
+    case jay::Type_Float32: stype = SType::FLOAT32; break;
+    case jay::Type_Float64: stype = SType::FLOAT64; break;
+    case jay::Type_Str32:   stype = SType::STR32; break;
+    case jay::Type_Str64:   stype = SType::STR64; break;
   }
 
+  Column* col = nullptr;
   MemoryRange databuf = extract_buffer(jaybuf, jcol->data());
-  if (jtype == jay::Type_Str32 || jtype == jay::Type_Str64) {
+  if (stype == SType::STR32 || stype == SType::STR64) {
     MemoryRange strbuf = extract_buffer(jaybuf, jcol->strdata());
-    col->replace_buffer(std::move(databuf), std::move(strbuf));
+    col = new_string_column(nrows, std::move(databuf), std::move(strbuf));
   } else {
-    col->replace_buffer(std::move(databuf));
+    col = Column::new_mbuf_column(stype, std::move(databuf));
   }
 
   Stats* stats = col->get_stats();

--- a/c/py_buffers.cc
+++ b/c/py_buffers.cc
@@ -175,8 +175,7 @@ static Column* convert_fwchararray_to_column(Py_buffer* view)
   }
 
   strbuf.resize(static_cast<size_t>(offset));
-  return new StringColumn<uint32_t>(nrows,
-                                    std::move(offbuf), std::move(strbuf));
+  return new_string_column(nrows, std::move(offbuf), std::move(strbuf));
 }
 
 
@@ -244,7 +243,7 @@ static Column* try_to_resolve_object_column(Column* col)
   xassert(offset < strbuf.size());
   strbuf.resize(offset);
   delete col;
-  return new StringColumn<uint32_t>(nrows, std::move(offbuf), std::move(strbuf));
+  return new_string_column(nrows, std::move(offbuf), std::move(strbuf));
 }
 
 

--- a/c/py_buffers.cc
+++ b/c/py_buffers.cc
@@ -224,7 +224,7 @@ static Column* try_to_resolve_object_column(Column* col)
   uint32_t offset = 0;
   for (size_t i = 0; i < nrows; ++i) {
     if (data[i] == Py_None) {
-      offsets[i] = offset | GETNA<uint32_t>();
+      offsets[i] = offset ^ GETNA<uint32_t>();
     } else {
       PyObject *z = PyUnicode_AsEncodedString(data[i], "utf-8", "strict");
       size_t sz = static_cast<size_t>(PyBytes_Size(z));

--- a/c/read/fread/fread_thread_context.cc
+++ b/c/read/fread/fread_thread_context.cc
@@ -277,7 +277,7 @@ void FreadThreadContext::postprocess() {
           coldata->str32.offset = output_offset;
         } else {
           xassert(coldata->str32.isna());
-          coldata->str32.offset = output_offset | GETNA<uint32_t>();
+          coldata->str32.offset = output_offset ^ GETNA<uint32_t>();
         }
         coldata += tbuf_ncols;
         xassert(output_offset <= sbuf.size());

--- a/c/wstringcol.cc
+++ b/c/wstringcol.cc
@@ -25,26 +25,59 @@ namespace dt {
 //------------------------------------------------------------------------------
 
 writable_string_col::writable_string_col(size_t nrows)
-  : strdata(new MemoryWritableBuffer(nrows)),
+  : strdata(nrows),
     offdata(MemoryRange::mem((nrows + 1) * sizeof(uint32_t))),
     n(nrows) {}
 
 
 Column* writable_string_col::to_column() && {
-  strdata->finalize();
+  strdata.finalize();
+  auto strbuf = strdata.get_mbuf();
   offdata.set_element<uint32_t>(0, 0);
-  return new StringColumn<uint32_t>(n, std::move(offdata), strdata->get_mbuf());
+  return new_string_column(n, std::move(offdata), std::move(strbuf));
+  /*
+  if (strbuf.size() <= MAX_STR32_BUFFER_SIZE && n <= MAX_STR32_NROWS) {
+    return new StringColumn<uint32_t>(n, std::move(offdata),
+                                      std::move(strbuf));
+  }
+  else {
+    // Offsets need to be recoded into uint64_t array
+    // TODO: this conversion can also be done in parallel, using
+    //
+    MemoryRange offsets64 = MemoryRange::mem(sizeof(uint64_t) * (n + 1));
+    auto data64 = static_cast<uint64_t*>(offsets64.xptr());
+    auto data32 = static_cast<const uint32_t*>(offdata.rptr());
+    uint64_t curr_offset = 0;
+    data64[0] = 0;
+    for (size_t i = 1; i <= n; ++i) {
+      uint32_t len = data32[i] - data32[i - 1];
+      if (len == GETNA<uint32_t>()) {
+        data64[i] = curr_offset | GETNA<uint64_t>();
+      } else {
+        curr_offset += len;
+        data64[i] = curr_offset;
+      }
+    }
+    xassert(curr_offset == strbuf.size());
+    return new StringColumn<uint64_t>(n, std::move(offsets64),
+                                      std::move(strbuf));
+  }
+  */
 }
 
+
+
+//------------------------------------------------------------------------------
+// writable_string_col::buffer
+//------------------------------------------------------------------------------
 
 writable_string_col::buffer::buffer(writable_string_col& s)
-  : col(s), strbuf(1024)
-{
-  strbuf_used = 0;
-  strbuf_write_pos = 0;
-  offptr = nullptr;
-  offptr0 = nullptr;
-}
+  : col(s),
+    strbuf(1024),
+    strbuf_used(0),
+    strbuf_write_pos(0),
+    offptr(nullptr),
+    offptr0(nullptr) {}
 
 
 void writable_string_col::buffer::write(const CString& str) {
@@ -56,13 +89,18 @@ void writable_string_col::buffer::write(const std::string& str) {
 }
 
 void writable_string_col::buffer::write(const char* ch, size_t len) {
+  // Note: `strbuf_used` may overflow during the conversion into uint32_t.
+  // If this happens, we would have to convert into STR64. It will be
+  // possible to restore the correct offsets, provided that no single string
+  // item exceeds MAX_ITEM_SIZE in size.
   if (ch) {
-    xassert(len < GETNA<uint32_t>());
+    xassert(len <= MAX_ITEM_SIZE);
     strbuf.ensuresize(strbuf_used + len);
     std::memcpy(strbuf.data() + strbuf_used, ch, len);
     strbuf_used += len;
     *offptr++ = static_cast<uint32_t>(strbuf_used);
   } else {
+    // Use XOR instead of OR in case the buffer overflows int32_t.
     *offptr++ = static_cast<uint32_t>(strbuf_used) ^ GETNA<uint32_t>();
   }
 }
@@ -73,12 +111,12 @@ void writable_string_col::buffer::write_na() {
 
 
 void writable_string_col::buffer::order() {
-  strbuf_write_pos = col.strdata->prep_write(strbuf_used, strbuf.data());
+  strbuf_write_pos = col.strdata.prep_write(strbuf_used, strbuf.data());
 }
 
 
 void writable_string_col::buffer::commit_and_start_new_chunk(size_t i0) {
-  col.strdata->write_at(strbuf_write_pos, strbuf_used, strbuf.data());
+  col.strdata.write_at(strbuf_write_pos, strbuf_used, strbuf.data());
   for (uint32_t* p = offptr0; p < offptr; ++p) {
     *p += strbuf_write_pos;
   }

--- a/c/wstringcol.cc
+++ b/c/wstringcol.cc
@@ -92,9 +92,9 @@ void writable_string_col::buffer::write(const char* ch, size_t len) {
   // Note: `strbuf_used` may overflow during the conversion into uint32_t.
   // If this happens, we would have to convert into STR64. It will be
   // possible to restore the correct offsets, provided that no single string
-  // item exceeds MAX_ITEM_SIZE in size.
+  // item exceeds MAX_STRING_SIZE in size.
   if (ch) {
-    xassert(len <= MAX_ITEM_SIZE);
+    xassert(len <= Column::MAX_STRING_SIZE);
     strbuf.ensuresize(strbuf_used + len);
     std::memcpy(strbuf.data() + strbuf_used, ch, len);
     strbuf_used += len;

--- a/c/wstringcol.cc
+++ b/c/wstringcol.cc
@@ -21,23 +21,23 @@ namespace dt {
 
 
 //------------------------------------------------------------------------------
-// fixed_height_string_col
+// writable_string_col
 //------------------------------------------------------------------------------
 
-fixed_height_string_col::fixed_height_string_col(size_t nrows)
+writable_string_col::writable_string_col(size_t nrows)
   : strdata(new MemoryWritableBuffer(nrows)),
     offdata(MemoryRange::mem((nrows + 1) * sizeof(uint32_t))),
     n(nrows) {}
 
 
-Column* fixed_height_string_col::to_column() && {
+Column* writable_string_col::to_column() && {
   strdata->finalize();
   offdata.set_element<uint32_t>(0, 0);
   return new StringColumn<uint32_t>(n, std::move(offdata), strdata->get_mbuf());
 }
 
 
-fixed_height_string_col::buffer::buffer(fixed_height_string_col& s)
+writable_string_col::buffer::buffer(writable_string_col& s)
   : col(s), strbuf(1024)
 {
   strbuf_used = 0;
@@ -47,36 +47,37 @@ fixed_height_string_col::buffer::buffer(fixed_height_string_col& s)
 }
 
 
-void fixed_height_string_col::buffer::write(const CString& str) {
+void writable_string_col::buffer::write(const CString& str) {
   write(str.ch, static_cast<size_t>(str.size));
 }
 
-void fixed_height_string_col::buffer::write(const std::string& str) {
+void writable_string_col::buffer::write(const std::string& str) {
   write(str.data(), str.size());
 }
 
-void fixed_height_string_col::buffer::write(const char* ch, size_t len) {
+void writable_string_col::buffer::write(const char* ch, size_t len) {
   if (ch) {
+    xassert(len < GETNA<uint32_t>());
     strbuf.ensuresize(strbuf_used + len);
     std::memcpy(strbuf.data() + strbuf_used, ch, len);
     strbuf_used += len;
     *offptr++ = static_cast<uint32_t>(strbuf_used);
   } else {
-    *offptr++ = static_cast<uint32_t>(strbuf_used) | GETNA<uint32_t>();
+    *offptr++ = static_cast<uint32_t>(strbuf_used) ^ GETNA<uint32_t>();
   }
 }
 
-void fixed_height_string_col::buffer::write_na() {
-  *offptr++ = static_cast<uint32_t>(strbuf_used) | GETNA<uint32_t>();
+void writable_string_col::buffer::write_na() {
+  *offptr++ = static_cast<uint32_t>(strbuf_used) ^ GETNA<uint32_t>();
 }
 
 
-void fixed_height_string_col::buffer::order() {
+void writable_string_col::buffer::order() {
   strbuf_write_pos = col.strdata->prep_write(strbuf_used, strbuf.data());
 }
 
 
-void fixed_height_string_col::buffer::commit_and_start_new_chunk(size_t i0) {
+void writable_string_col::buffer::commit_and_start_new_chunk(size_t i0) {
   col.strdata->write_at(strbuf_write_pos, strbuf_used, strbuf.data());
   for (uint32_t* p = offptr0; p < offptr; ++p) {
     *p += strbuf_write_pos;

--- a/c/wstringcol.cc
+++ b/c/wstringcol.cc
@@ -35,34 +35,6 @@ Column* writable_string_col::to_column() && {
   auto strbuf = strdata.get_mbuf();
   offdata.set_element<uint32_t>(0, 0);
   return new_string_column(n, std::move(offdata), std::move(strbuf));
-  /*
-  if (strbuf.size() <= MAX_STR32_BUFFER_SIZE && n <= MAX_STR32_NROWS) {
-    return new StringColumn<uint32_t>(n, std::move(offdata),
-                                      std::move(strbuf));
-  }
-  else {
-    // Offsets need to be recoded into uint64_t array
-    // TODO: this conversion can also be done in parallel, using
-    //
-    MemoryRange offsets64 = MemoryRange::mem(sizeof(uint64_t) * (n + 1));
-    auto data64 = static_cast<uint64_t*>(offsets64.xptr());
-    auto data32 = static_cast<const uint32_t*>(offdata.rptr());
-    uint64_t curr_offset = 0;
-    data64[0] = 0;
-    for (size_t i = 1; i <= n; ++i) {
-      uint32_t len = data32[i] - data32[i - 1];
-      if (len == GETNA<uint32_t>()) {
-        data64[i] = curr_offset | GETNA<uint64_t>();
-      } else {
-        curr_offset += len;
-        data64[i] = curr_offset;
-      }
-    }
-    xassert(curr_offset == strbuf.size());
-    return new StringColumn<uint64_t>(n, std::move(offsets64),
-                                      std::move(strbuf));
-  }
-  */
 }
 
 
@@ -89,10 +61,6 @@ void writable_string_col::buffer::write(const std::string& str) {
 }
 
 void writable_string_col::buffer::write(const char* ch, size_t len) {
-  // Note: `strbuf_used` may overflow during the conversion into uint32_t.
-  // If this happens, we would have to convert into STR64. It will be
-  // possible to restore the correct offsets, provided that no single string
-  // item exceeds MAX_STRING_SIZE in size.
   if (ch) {
     xassert(len <= Column::MAX_STRING_SIZE);
     strbuf.ensuresize(strbuf_used + len);

--- a/c/wstringcol.h
+++ b/c/wstringcol.h
@@ -29,7 +29,7 @@ namespace dt {
 // TODO: merge into parallel.h
 
 //------------------------------------------------------------------------------
-// Fixed-height writable string column
+// (Fixed-height) writable string column
 //------------------------------------------------------------------------------
 
 /**
@@ -39,9 +39,13 @@ namespace dt {
  */
 class writable_string_col {
   private:
-    std::unique_ptr<MemoryWritableBuffer> strdata;
+    MemoryWritableBuffer strdata;
     MemoryRange offdata;
     size_t n;
+
+    static constexpr size_t MAX_ITEM_SIZE = 0x7FFFFFFF;
+    static constexpr size_t MAX_STR32_BUFFER_SIZE = 0x7FFFFFFF;
+    static constexpr size_t MAX_STR32_NROWS = 0x7FFFFFFF;
 
   public:
     writable_string_col(size_t nrows);

--- a/c/wstringcol.h
+++ b/c/wstringcol.h
@@ -26,6 +26,7 @@ template <typename T> class StringColumn;
 
 namespace dt {
 
+// TODO: merge into parallel.h
 
 //------------------------------------------------------------------------------
 // Fixed-height writable string column
@@ -36,19 +37,19 @@ namespace dt {
  * number of rows is known in advance).
  *
  */
-class fixed_height_string_col {
+class writable_string_col {
   private:
     std::unique_ptr<MemoryWritableBuffer> strdata;
     MemoryRange offdata;
     size_t n;
 
   public:
-    fixed_height_string_col(size_t nrows);
+    writable_string_col(size_t nrows);
     Column* to_column() &&;
 
     class buffer {
       private:
-        fixed_height_string_col& col;
+        writable_string_col& col;
         dt::array<char> strbuf;
         size_t strbuf_used;
         size_t strbuf_write_pos;
@@ -56,7 +57,7 @@ class fixed_height_string_col {
         uint32_t* offptr0;
 
       public:
-        buffer(fixed_height_string_col&);
+        buffer(writable_string_col&);
 
         void write(const CString&);
         void write(const std::string&);
@@ -69,7 +70,7 @@ class fixed_height_string_col {
 };
 
 
-
+using fixed_height_string_col = writable_string_col;
 
 
 

--- a/c/wstringcol.h
+++ b/c/wstringcol.h
@@ -43,10 +43,6 @@ class writable_string_col {
     MemoryRange offdata;
     size_t n;
 
-    static constexpr size_t MAX_ITEM_SIZE = 0x7FFFFFFF;
-    static constexpr size_t MAX_STR32_BUFFER_SIZE = 0x7FFFFFFF;
-    static constexpr size_t MAX_STR32_NROWS = 0x7FFFFFFF;
-
   public:
     writable_string_col(size_t nrows);
     Column* to_column() &&;

--- a/tests/munging/test_replace.py
+++ b/tests/munging/test_replace.py
@@ -252,6 +252,26 @@ def test_replace_str_large(seed):
     assert df.to_list() == [src]
 
 
+def test_replace_str_huge():
+    # Issue #1694: the output column is too large to fit into str32, so it
+    # must be converted into str64.
+    n = 10000000
+    src = ["a"] * n
+    src[-1] = None
+    src[-100] = None
+    src[-10000] = None
+    src[-1000000] = None
+    DT = dt.Frame(src)
+    assert DT.stypes == (dt.str32,)
+    DT.replace("a", "A" * 250)
+    DT.internal.check()
+    assert DT.stypes == (dt.str64,)
+    assert DT.shape == (n, 1)
+    assert DT[-2, 0] == "A" * 250
+    assert DT[-1, 0] is None
+    assert DT[-100, 0] is None
+
+
 
 
 #-------------------------------------------------------------------------------

